### PR TITLE
Fix for excess IP release race condition | handshake between agent and operator

### LINKED ISSUE #13412: AWS ENI: Pod assigned IP that is not attached to an ENI
Two pods on my cluster are geting an IP that is not attached to an ENI

Cilium version is 1.8.4
EKS Kubernetes version 1.16

Observed this issue after upgrading from 1.8.3 to 1.8.4 and adding nodes to the cluster

This output shows a CiliumNode where the rbac-manager pod is assigned the IP  10.80.17.39 and that IP is not associated with any of the active ENIs 

```
apiVersion: cilium.io/v2
kind: CiliumNode
metadata:
  creationTimestamp: "2020-10-05T15:17:10Z"
  generation: 19
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: m5.large
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east-1
    failure-domain.beta.kubernetes.io/zone: us-east-1a
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ip-10-80-17-191.us-east-1.compute.internal
    kubernetes.io/os: linux
    nite.tradestation.io/role: worker
  name: ip-10-80-17-191.ec2.internal
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: ip-10-80-17-191.ec2.internal
    uid: cf85a4d5-8d93-4ba4-88d5-9ac8251df556
  resourceVersion: "7640153"
  selfLink: /apis/cilium.io/v2/ciliumnodes/ip-10-80-17-191.ec2.internal
  uid: 1b6030a4-eb02-4634-8410-a3e1f9051103
spec:
  addresses:
  - ip: 10.80.17.191
    type: InternalIP
  - ip: 10.80.17.232
    type: CiliumInternalIP
  azure: {}
  encryption: {}
  eni:
    availability-zone: us-east-1a
    delete-on-termination: true
    first-interface-index: 1
    instance-type: m5.large
    subnet-tags:
      NetworkType: private
    vpc-id: vpc-0e030555a019aefa4
  health:
    ipv4: 10.80.17.16
  instance-id: i-0466c2a16aaa4eb42
  ipam:
    podCIDRs:
    - 10.191.0.0/16
    pool:
      10.80.17.16:
        resource: eni-019ecf4a65359906b
      10.80.17.24:
        resource: eni-019ecf4a65359906b
      10.80.17.28:
        resource: eni-019ecf4a65359906b
      10.80.17.30:
        resource: eni-0d938326ff46f5e20
      10.80.17.85:
        resource: eni-0d938326ff46f5e20
      10.80.17.86:
        resource: eni-0d938326ff46f5e20
      10.80.17.100:
        resource: eni-0d938326ff46f5e20
      10.80.17.102:
        resource: eni-019ecf4a65359906b
      10.80.17.108:
        resource: eni-0d938326ff46f5e20
      10.80.17.112:
        resource: eni-019ecf4a65359906b
      10.80.17.116:
        resource: eni-0d938326ff46f5e20
      10.80.17.136:
        resource: eni-019ecf4a65359906b
      10.80.17.186:
        resource: eni-019ecf4a65359906b
      10.80.17.207:
        resource: eni-019ecf4a65359906b
      10.80.17.232:
        resource: eni-0d938326ff46f5e20
      10.80.17.234:
        resource: eni-019ecf4a65359906b
      10.80.17.239:
        resource: eni-0d938326ff46f5e20
      10.80.17.251:
        resource: eni-0d938326ff46f5e20
      10.80.17.253:
        resource: eni-019ecf4a65359906b
    pre-allocate: 8
status:
  azure: {}
  eni:
    enis:
      eni-019ecf4a65359906b:
        addresses:
        - 10.80.17.16
        - 10.80.17.253
        - 10.80.17.136
        - 10.80.17.234
        - 10.80.17.186
        - 10.80.17.28
        - 10.80.17.24
        - 10.80.17.207
        - 10.80.17.102
        - 10.80.17.112
        description: Cilium-CNI (i-0466c2a16aaa4eb42)
        id: eni-019ecf4a65359906b
        ip: 10.80.17.16
        mac: 02:30:e9:1e:a2:79
        number: 1
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
      eni-0cd1b87a2f1c93097:
        addresses:
        - 10.80.17.191
        id: eni-0cd1b87a2f1c93097
        ip: 10.80.17.191
        mac: 02:1b:da:9c:0e:51
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
      eni-0d938326ff46f5e20:
        addresses:
        - 10.80.17.116
        - 10.80.17.232
        - 10.80.17.251
        - 10.80.17.239
        - 10.80.17.100
        - 10.80.17.86
        - 10.80.17.108
        - 10.80.17.30
        - 10.80.17.85
        description: Cilium-CNI (i-0466c2a16aaa4eb42)
        id: eni-0d938326ff46f5e20
        ip: 10.80.17.116
        mac: 02:9f:02:c6:08:8d
        number: 2
        security-groups:
        - sg-0a521e746d55bca88
        subnet:
          cidr: 10.80.17.0/24
          id: subnet-0ec1716995dd5c73b
        vpc:
          id: vpc-0e030555a019aefa4
          primary-cidr: 10.80.16.0/22
  ipam:
    operator-status: {}
    used:
      10.80.17.16:
        owner: health
        resource: eni-019ecf4a65359906b
      10.80.17.28:
        owner: kube-system/efs-provisioner-gp-5b488cf9bf-pwcvq
        resource: eni-019ecf4a65359906b
      10.80.17.39:
        owner: kube-system/rbac-manager-5f56b4d9c8-lkvxw
        resource: eni-0d938326ff46f5e20
      10.80.17.100:
        owner: kube-system/hubble-relay-56dcd9c4b-pmcrm
        resource: eni-0d938326ff46f5e20
      10.80.17.116:
        owner: kube-system/fluentd-xmtsc
        resource: eni-0d938326ff46f5e20
      10.80.17.136:
        owner: default/pod-to-a-denied-cnp-6967cb6f7f-qt4zx
        resource: eni-019ecf4a65359906b
      10.80.17.232:
        owner: router
        resource: eni-0d938326ff46f5e20
      10.80.17.234:
        owner: nite-opa/gatekeeper-controller-manager-85c6977484-4vjkv
        resource: eni-019ecf4a65359906b
      10.80.17.239:
        owner: kube-system/prometheus-prometheus-prometheus-oper-prometheus-0
        resource: eni-0d938326ff46f5e20
      10.80.17.251:
        owner: kube-system/datadog-c82cg
        resource: eni-0d938326ff46f5e20
      10.80.17.253:
        owner: default/pod-to-b-multi-node-clusterip-6fcdb496c7-vsn68
        resource: eni-019ecf4a65359906b
```

### PR DESCRIPTION
**Pull request by hemanthmalla**
Forward port of https://github.com/DataDog/cilium/pull/24

Currently there's a 15 sec delay in updating ciliumnode (CN) CRD after IP allocation to a pod, meanwhile the operator can determine that a node has excess IPs and release the IP causing pod connectivity issues.

This PR introduces a handshake between the operator and agent before releasing an excess IP to avoid the race condition. A new operator flag `excess-ip-release-delay` is added to control how long operator should wait before considering an IP for release (defaults to 180 secs). This is done to better handle IP reuse during rollouts. Operator and agent use a new map in cilium node status `.status.ipam.release-ips` to exchange information during the handshake.

Fixes: #13412

handshake_upstream

The following alternative options were considered : 

* We could force a CN write to go through before allocation but it can result in too many writes from the agent.

* The operator picks the ENI with the most free IPs to release excess IPs from, this is done so that the ENI could potentially be released in the future. We could move the excess detection logic to the agent, but this involves calling into cloud provider specific implementation, which AFAIU the agent IPAM implementation does not.

```release-note
Fix for excess IP release race condition. New operator flag excess-ip-release-delay is introduced to control waiting period before marking an IP for release.
```
Note : https://github.com/DataDog/cilium/pull/24 already received some valuable feedback from @aanm @gandro and @christarazi

### CODE REVIEW

File: pkg/ipam/crd.go [hunk 1]
```diff
@@ -323,6 +323,54 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
...
+		}
+		// Retrieve the appropriate allocator
+		var allocator *crdAllocator
+		var ipFamily Family
+		if ipAddr := net.ParseIP(ip); ipAddr != nil {
+			if ipAddr.To4() != nil {
+				ipFamily = IPv4
+			} else {
+				ipFamily = IPv6
+			}
```
Comment (gandro) [inline pkg/ipam/crd.go]: tiny nit (feel free to ignore): 
```suggestion
			ipFamily = DeriveFamiliy(ipAddr)
```

File: pkg/ipam/crd.go [hunk 2]
```diff
@@ -389,6 +437,12 @@ func (n *nodeStore) allocate(ip net.IP) (*ipamTypes.AllocationIP, error) {
 		return nil, fmt.Errorf("No IPs available")
 	}
 
+	if status, ok := n.ownNode.Status.IPAM.ReleaseIPs[ip.String()]; ok {
+		if status == ipamOption.IPAMMarkForRelease || status == ipamOption.IPAMReadyForRelease {
+			return nil, fmt.Errorf("IP not available, marked or ready for release")
+		}
+	}
```
Comment (gandro) [inline pkg/ipam/crd.go]: Sorry I didn't realize this last time, but a check like this also needs to be added to `nodeStore.allocateNext()`. Otherwise the IP will be handed out, as IPs are usually allocated via `allocateNext()`.

i.e. adding a check like this:
```go
	if status, ok := n.ownNode.Status.IPAM.ReleaseIPs[ip]; ok {
		if status == ipamOption.IPAMMarkForRelease || status == ipamOption.IPAMReadyForRelease {
			continue // IP not available
		}
	}
```

to the iteration here:
https://github.com/cilium/cilium/blob/a10a784bfba4eae0651a7387de95216fb3ec4223/pkg/ipam/crd.go#L465

In addition, I think we also need to add a similar check to `hasMinimumIPsInPool`.

File: pkg/ipam/types/types.go [hunk 3]
```diff
@@ -114,6 +114,15 @@ type IPAMStatus struct {
...
 	OperatorStatus OperatorStatus `json:"operator-status,omitempty"`
+
+	// ReleaseIPs tracks the state for every IP considered for release.
+	// value can be one of the following string :
+	// * marked-for-release : Set by operator as possible candidate for IP
+	// * ready-for-release  : Acknowledged as safe to release by agent
+	// * do-not-release     : IP already in use / not owned by the node. Set by agent
+	//
+	// +optional
+	ReleaseIPs map[string]string `json:"release-ips,omitempty"`
```
Comment (gandro) [inline pkg/ipam/types/types.go]: nit: You could introduce a newtype with K8s validation requirements here. For example:

```go
// +kubebuilder:validation:Enum=marked-for-release;ready-for-release;do-not-release
type IPReleaseStatus string
```

And then change this type to:

```suggestion
	ReleaseIPs map[string]IPReleaseStatus `json:"release-ips,omitempty"`
```

**Review by gandro — CHANGES_REQUESTED** [review]
Awesome, thanks a ton for addressing this long-standing issue.

I think we need to ensure that `allocateNext` doesn't hand out IPs which are marked for release too (see inline comment), the rest are smaller nits.

File: pkg/ipam/crd.go (hunk 1 above)
Comment (hemanthmalla) [reply pkg/ipam/crd.go]: Updated this.

File: pkg/ipam/crd.go (hunk 2 above)
Comment (hemanthmalla) [reply pkg/ipam/crd.go]: Added checks to allocateNext() and updated `hasMinimumIPsInPool` to factor in IPs currently in handshake.

File: pkg/ipam/types/types.go (hunk 3 above)
Comment (hemanthmalla) [reply pkg/ipam/types/types.go]: Added the new type and validations. Had to update some tests to convert from `IPReleaseStatus ` to string

**Comment by hemanthmalla** [comment]
There's one more corner case that needs addressing.

After the handshake is complete and the operator is done releasing an IP, CiliumNode status (release-ips) and spec (pool) are updated in two consecutive requests. There's a tiny window between the two updates where the entry is removed from `.status.ipam.release-ips` but the IP is till present in `spec.ipam.pool`. It might be possible that the IP can be allocated between these requests.

Can i flip the order of update only when there are changes to IP release status ? Comments in `syncToAPIServer()` say that the status should always be updated first though. I don't fully understand why. 
https://github.com/cilium/cilium/blob/eeb7f1b464b283dc0c2b2bcada2a149b40372437/pkg/ipam/node.go#L700-L701

**Comment by gandro** [comment]
> There's one more corner case that needs addressing.
> 
> After the handshake is complete and the operator is done releasing an IP, CiliumNode status (release-ips) and spec (pool) are updated in two consecutive requests. There's a tiny window between the two updates where the entry is removed from `.status.ipam.release-ips` but the IP is till present in `spec.ipam.pool`. It might be possible that the IP can be allocated between these requests.

Oh wow, good catch! In dynamic cluster pool, we avoid this by having the agent remove the entry from `status` once it is no longer present in `pool`, which might also be an option here.

> @gandro / @aanm Can i flip the order of update only when there are changes to IP release status ? Comments in `syncToAPIServer()` say that the status should always be updated first though. I don't fully understand why.

While I'm also not sure, I _believe_ the order might have to do with the fact that we need to know the ENI metadata for each IP present in spec. If the agent sees an IP in the pool before it knows about its ENI, on other words to avoid this error being hit:
https://github.com/cilium/cilium/blob/124fa2be77c85d6bce65c15bbd91f69507876f70/pkg/ipam/crd.go#L528

**Comment by gandro** [comment]
The two CI failures (besides checkpatch) look unrelated:

- Smoke Test with IPv6 has a failing `host-to-b-multi-node-clusterip` pod, which looks like a variant of https://github.com/cilium/cilium/issues/12279 https://github.com/cilium/cilium/runs/4305950192?check_suite_focus=true
- test-1.22-4.9 hit https://github.com/cilium/cilium/issues/17919 and #17881

**Review by gandro — COMMENTED** [review]
The implementation looks good to me now! Thanks al lot!

There seems to be a unit test failure that needs to be fixed. Let me know if you need any help with that.

Edit: https://app.travis-ci.com/github/cilium/cilium/jobs/549921704

**Review by gandro — APPROVED** [review]
Discussed offline. Looks like the previous Travis failure was actually unrelated and a case of the flaky unit tests https://github.com/cilium/cilium/issues/11560

Looks good to me now! :rocket:

**Review by gandro — APPROVED** [review]

File: operator/provider_aws_flags.go [hunk 4]
```diff
@@ -29,6 +29,9 @@ func init() {
 	flags.Bool(operatorOption.AWSReleaseExcessIPs, false, "Enable releasing excess free IP addresses from AWS ENI.")
 	option.BindEnv(operatorOption.AWSReleaseExcessIPs)
 
+	flags.Int(operatorOption.ExcessIPReleaseDelay, 180, "Number of seconds operator would wait before it releases an IP previously marked as excess. Defaults to 180")
+	option.BindEnv(operatorOption.ExcessIPReleaseDelay)
+
```
Comment (christarazi) [inline operator/provider_aws_flags.go]: I think the `flags` package will actually already show the default value in the text string, so no need to add it yourself.

File: pkg/ipam/crd.go [hunk 5]
```diff
@@ -330,6 +330,54 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
...
+		allocator.mutex.Lock()
+		if _, ok := allocator.allocated[ip]; ok {
+			// IP still in use, update the operator to stop releasing the IP.
+			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMDoNotRelease
+		} else {
+			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMReadyForRelease
+		}
+		allocator.mutex.Unlock()
+		releaseUpstreamSyncNeeded = true
+	}
```
Comment (christarazi) [inline pkg/ipam/crd.go]: Would it make sense to move this code to a isolated function and have it return `releaseUpstreamSyncNeeded`?

File: pkg/ipam/node.go [hunk 6]
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
...
+			if markedIP == ip {
+				stillMarkedForRelease = true
+				break
+			}
+		}
+		if !stillMarkedForRelease {
+			// n.determineMaintenanceAction() only returns the IPs on the interface with maximum number of IPs that
+			// can be freed up. If the selected interface changes or if this IP is not excess anymore, remove entry
+			// from local map and cilium node status if n.ipReleaseStatus is already updated.
+			// from n.ipReleaseStatus
```
Comment (christarazi) [inline pkg/ipam/node.go]: Is the second to last sentence supossed to end or is `from n.ipReleaseStatus` part of it?

File: pkg/ipam/node.go [hunk 7]
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
...
+					delete(n.ipReleaseStatus, markedIP)
+				}
+				continue
+			}
+		}
+		ipsToMark = append(ipsToMark, markedIP)
+	}
+	// Update cilium node CRD with IPs that need to be marked for release
+	for _, ip := range ipsToMark {
+		scopedLog.WithFields(logrus.Fields{"ip": ip}).Debug("Marking IP for release")
```
Comment (christarazi) [inline pkg/ipam/node.go]: logfields.IPAddr

File: pkg/ipam/node.go [hunk 8]
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
...
+	if len(ipsToRelease) > 0 {
+		a.release.IPsToRelease = ipsToRelease
+		scopedLog = scopedLog.WithFields(logrus.Fields{
+			"available":         n.stats.AvailableIPs,
+			"used":              n.stats.UsedIPs,
+			"excess":            n.stats.ExcessIPs,
+			"excessIps":         a.release.IPsToRelease,
+			"releasing":         ipsToRelease,
+			"selectedInterface": a.release.InterfaceID,
+			"selectedPoolID":    a.release.PoolID,
```
Comment (christarazi) [inline pkg/ipam/node.go:682]: Ideally, since these are a part of an INFO level log, they should be inside the logfields package. However, there are other parts of the code that also suffer the same thing, so let's just followup on this later, doesn't have to be part of this PR.

File: pkg/ipam/crd.go [hunk 9]
```diff
@@ -346,11 +349,7 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
 		var allocator *crdAllocator
 		var ipFamily Family
 		if ipAddr := net.ParseIP(ip); ipAddr != nil {
-			if ipAddr.To4() != nil {
-				ipFamily = IPv4
-			} else {
-				ipFamily = IPv6
-			}
+			ipFamily = DeriveFamily(ipAddr)
```
Comment (christarazi) [inline pkg/ipam/crd.go]: This could be part of the first commit, no?

File: pkg/ipam/crd.go [hunk 10]
```diff
@@ -337,12 +337,20 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
 	releaseUpstreamSyncNeeded := false
 	// ACK or NACK IPs marked for release by the operator
 	for ip, status := range n.ownNode.Status.IPAM.ReleaseIPs {
-		if status != ipamOption.IPAMMarkForRelease || n.ownNode.Spec.IPAM.Pool == nil {
```
Comment (christarazi) [inline pkg/ipam/crd.go]: nit on commit msg title: We could probably just shorten it to 

```
Fix CN update order by adding new `released` IPAM state
```

**Review by christarazi — APPROVED** [review]
Thanks for forward porting! The changes LGTM overall. I am marking my review as approved, but there are suggestions that I'd like to see followed up on in a later PR, so that we can get this PR in. Feel free to address my comments as you see fit in this PR however. After this is merged, we can create a tracking issue and put all the comments that need following up on in there. Please use "resolve conversion" for comments that you've already addressed in this PR.

Note, for any suggestions related to commits, could you address those in this PR as those can't be changed in a followup?

File: operator/provider_aws_flags.go (hunk 4 above)
Comment (hemanthmalla) [reply operator/provider_aws_flags.go]: Removed the default part.

File: pkg/ipam/node.go (hunk 6 above)
Comment (hemanthmalla) [reply pkg/ipam/node.go]: This was a stale comment ( probably from applying patch). Fixed this.

File: pkg/ipam/crd.go (hunk 9 above)
Comment (hemanthmalla) [reply pkg/ipam/crd.go]: The first and second commits are now squashed.

File: pkg/ipam/crd.go (hunk 10 above)
Comment (hemanthmalla) [reply pkg/ipam/crd.go]: Updated the commit message.

**Comment by hemanthmalla** [comment]
@christarazi Thanks for the review. Addressed most the the feedback except the conversations still unresolved. There's also one more refactor suggestion in the original PR https://github.com/DataDog/cilium/pull/24#discussion_r740761791 we can add to the list for the future PR.

**Comment by aanm** [comment]
@hemanthmalla the unit tests failures are legitimate

**Comment by nbusseneau** [comment]
Thanks @hemanthmalla! Merging.

File: pkg/aws/eni/node.go [hunk 11]
```diff
@@ -119,9 +120,17 @@ func (n *Node) PrepareIPRelease(excessIPs int, scopedLog *logrus.Entry) *ipam.Re
 	n.mutex.Lock()
 	defer n.mutex.Unlock()
 
+	// Needed for selecting the same ENI to release IPs from
+	// when more than one ENI qualifies for release
+	eniIds := make([]string, 0, len(n.enis))
+	for k := range n.enis {
+		eniIds = append(eniIds, k)
+	}
+	sort.Strings(eniIds)
```
Comment (wu0407) [inline pkg/aws/eni/node.go:129]: Same iterate order not ensure selecting the same ENI to release IPs in multi `PrepareIPRelease` called. what is use case ? @hemanthmalla

Comment (wu0407) [reply pkg/aws/eni/node.go:129]: Because select the ENI with the most addresses available for release  in every call `PrepareIPRelease`
